### PR TITLE
Fix 'getConsoleOutput' + add tests

### LINKED ISSUE #224: Latest released version does not properly display logs
### Jenkins and plugins versions report

```
Jenkins: 2.375.3
OS: Linux - 5.10.147+
Java: 11.0.18 - Eclipse Adoptium (OpenJDK 64-Bit Server VM)
---
ansicolor:1.0.2
antisamy-markup-formatter:155.v795fb_8702324
apache-httpcomponents-client-4-api:4.5.13-138.v4e7d9a_7b_a_e61
authentication-tokens:1.4
authorize-project:1.4.0
basic-branch-build-strategies:71.vc1421f89888e
blueocean:1.27.1
blueocean-autofavorite:1.2.5
blueocean-bitbucket-pipeline:1.27.1
blueocean-commons:1.27.1
blueocean-config:1.27.1
blueocean-core-js:1.27.1
blueocean-dashboard:1.27.1
blueocean-display-url:2.4.1
blueocean-events:1.27.1
blueocean-git-pipeline:1.27.1
blueocean-github-pipeline:1.27.1
blueocean-i18n:1.27.1
blueocean-jwt:1.27.1
blueocean-personalization:1.27.1
blueocean-pipeline-api-impl:1.27.1
blueocean-pipeline-editor:1.27.1
blueocean-pipeline-scm-api:1.27.1
blueocean-rest:1.27.1
blueocean-rest-impl:1.27.1
blueocean-web:1.27.1
bootstrap5-api:5.2.1-3
bouncycastle-api:2.26
branch-api:2.1051.v9985666b_f6cc
build-with-parameters:1.6
caffeine-api:2.9.3-65.v6a_47d0f4d1fe
checks-api:1.8.1
cloudbees-bitbucket-branch-source:796.v6cb_1559e1673
cloudbees-folder:6.795.v3e23d3c6f194
command-launcher:90.v669d7ccb_7c31
commons-lang3-api:3.12.0-36.vd97de6465d5b_
commons-text-api:1.10.0-27.vb_fa_3896786a_7
conditional-buildstep:1.4.2
configuration-as-code:1569.vb_72405b_80249
credentials:1214.v1de940103927
credentials-binding:523.vd859a_4b_122e6
display-url-api:2.3.7
docker-commons:1.21
docker-workflow:563.vd5d2e5c4007f
durable-task:504.vb10d1ae5ba2f
echarts-api:5.4.0-1
envinject:2.881.v37c62073ff97
envinject-api:1.199.v3ce31253ed13
favorite:2.4.1
font-awesome-api:6.2.0-3
gcp-secrets-manager-credentials-provider:0.3.1
ghprb:1.42.2
git:5.0.0
git-client:4.1.0
git-parameter:0.9.18
github:1.37.0
github-api:1.303-400.v35c2d8258028
github-branch-source:1701.v00cc8184df93
handy-uri-templates-2-api:2.1.8-22.v77d5b_75e6953
htmlpublisher:1.31
instance-identity:142.v04572ca_5b_265
ionicons-api:31.v4757b_6987003
jackson2-api:2.13.4.20221013-295.v8e29ea_354141
jakarta-activation-api:2.0.1-3
jakarta-mail-api:2.0.1-3
javax-activation-api:1.2.0-5
javax-mail-api:1.6.2-8
jaxb:2.3.7-1
jdk-tool:63.v62d2fd4b_4793
jenkins-design-language:1.27.1
jjwt-api:0.11.5-77.v646c772fddb_0
job-dsl:1.81
jquery3-api:3.6.1-2
junit:1177.v90374a_ef4d09
kubernetes:3883.v4d70a_a_a_df034
kubernetes-client-api:6.4.1-208.vfe09a_9362c2c
kubernetes-credentials:0.10.0
mailer:438.v02c7f0a_12fa_4
matrix-auth:3.1.5
matrix-project:785.v06b_7f47b_c631
metrics:4.2.13-420.vea_2f17932dd6
mina-sshd-api-common:2.9.2-50.va_0e1f42659a_a
mina-sshd-api-core:2.9.2-50.va_0e1f42659a_a
okhttp-api:4.9.3-108.v0feda04578cf
parameterized-trigger:2.45
pipeline-build-step:2.18.1
pipeline-graph-analysis:202.va_d268e64deb_3
pipeline-graph-view:157.vd6123e1b_a_e2e
pipeline-groovy-lib:629.vb_5627b_ee2104
pipeline-input-step:466.v6d0a_5df34f81
pipeline-milestone-step:111.v449306f708b_7
pipeline-model-api:2.2118.v31fd5b_9944b_5
pipeline-model-definition:2.2118.v31fd5b_9944b_5
pipeline-model-extensions:2.2118.v31fd5b_9944b_5
pipeline-rest-api:2.31
pipeline-stage-step:305.ve96d0205c1c6
pipeline-stage-tags-metadata:2.2118.v31fd5b_9944b_5
pipeline-stage-view:2.31
plain-credentials:143.v1b_df8b_d3b_e48
plugin-usage-plugin:4.0
plugin-util-api:2.20.0
popper2-api:2.11.6-2
prometheus:2.1.1
pubsub-light:1.17
rebuild:1.34
role-strategy:562.v44e9a_e828d0e
run-condition:1.5
saml:4.372.v89f13e4c9e97
scm-api:631.v9143df5b_e4a_a
script-security:1229.v4880b_b_e905a_6
snakeyaml-api:1.33-90.v80dcb_3814d35
sse-gateway:1.26
ssh-agent:295.v9ca_a_1c7cc3a_a_
ssh-credentials:305.v8f4381501156
sshd:3.249.v2dc2ea_416e33
structs:324.va_f5d6774f3a_d
timestamper:1.21
token-macro:321.vd7cc1f2a_52c8
trilead-api:2.84.v72119de229b_7
uno-choice:2.6.4
variant:59.vf075fe829ccb
workflow-api:1208.v0cc7c6e0da_9e
workflow-basic-steps:1010.vf7a_b_98e847c1
workflow-cps:3618.v13db_a_21f0fcf
workflow-durable-task-step:1234.v019404b_3832a
workflow-job:1268.v6eb_e2ee1a_85a
workflow-multibranch:733.v109046189126
workflow-scm-step:400.v6b_89a_1317c9a_
workflow-step-api:639.v6eca_cd8c04a_a_
workflow-support:839.v35e2736cfd5c
```

### What Operating System are you using (both controller, and any agents involved in the problem)?

Linux

### Reproduction steps

Jenkinsfile

```
pipeline {
    agent any
    stages {
        stage('Main') {
            steps {
                sh 'hostname'
                sh 'ls -lta'
                sh 'ls -lta /'
                sh 'sleep 60'
                sh 'ps -efww'
            }
        }
    }
}

```

### Expected Results

Blue Ocean:

Screenshot 2023-02-24 at 09 23 46

### Actual Results

Screenshot 2023-02-24 at 09 21 45

### Anything else?

_No response_

### LINKED ISSUE #228: Pipeline Graph doesn't show child stages for Scripted Pipeline
### Jenkins and plugins versions report

<details>
  <summary>Environment</summary>

  
  ```text
  Paste the output here
  ```

</details>

### What Operating System are you using (both controller, and any agents involved in the problem)?

Jenkins: 2.361.4
Pipeline Graph View Plugin: 157.vd6123e1b_a_e2e

### Reproduction steps

When using this scripted Pipeline I am expecting the graph to show a `World` stage as a sibling of `Hello`, but it doesn't get displayed at all.
```
stage("Hello") {
    stage("World") {
        echo("Hello, world!")
    }
}
```

### Expected Results

I expect the same graph as this declarative Pipeline:
```
pipeline {
    agent any

    stages {
        stage('Hello') {
            stages {
                stage('World') {
                    steps {
                        echo 'Hello World'   
                    }
                }
            }
        }
    }
}
```
Screenshot 2023-02-24 at 13 37 05

### Actual Results

I don't see the 'world' stage at all (this is also true for the Pipeline Console).
Screenshot 2023-02-24 at 13 47 36

### Anything else?

The error seems to be that nested scripted stages are skipped: 
https://github.com/jenkinsci/pipeline-graph-view-plugin/blob/main/src/main/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineNodeGraphVisitor.java#L221

This was introduced here: https://github.com/jenkinsci/blueocean-plugin/pull/1778  (maybe scripted Pipelines didn't support this in the past?)

When I remove this section I get the expected graph:
Screenshot 2023-02-24 at 13 53 19

### PR DESCRIPTION
**Pull request by timbrown5**
Fixes #224 + add tests

- 'getConsoleOutput' wasn't setting `text` before return.
- `getLogForNode` was always returning null (no early return). 
- fix logic that was setting 'startByte' - caused the 'Get more logs' button to appear in error.

Moved some code to make it easier to test.

- [x] Make sure you are opening from a **topic/feature/bugfix branch** (right side) and not your main branch!
- [x] Ensure that the pull request title represents the desired changelog entry
- [x] Please describe what you did
- [x] Link to relevant issues in GitHub or Jira
- [x] Ensure you have provided tests - that demonstrates feature works or fixes the issue

### CODE REVIEW

**Comment by timbrown5** [comment]
It would be nice to find a way to make the tests quicker (e.g. not needing a Jenkins job to run). Does anyone know if there's a way to serialise a run?

I feel there is a better way to get steps in the tests (shouldn't rely on PipelineStepVisitor as if it breaks a lot of other tests break). I tried with `getNodesByDisplayName` - which uses the upstream `FlowGraphTable`, but couldn't find the correct way to get a step (works fine for stages).

**Comment by timbrown5** [comment]
I think it fails due to windows line endings 🙄

File: src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java [hunk 1]
```diff
@@ -55,35 +55,44 @@ public String getIconClassName() {
...
+      return HttpResponses.okJSON(getSteps(nodeId));
     } else {
       return HttpResponses.errorJSON("Error getting console text");
     }
   }
 
+  // Private method for testing.
+  protected JSONObject getSteps(String nodeId) throws IOException {
+    logger.debug("getSteps was passed nodeId '" + nodeId + "'.");
+    ObjectMapper mapper = new ObjectMapper();
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java:67]: object mappers should generally just be initialised once and then used from a static variable. I see you've just moved it here so non-blocking.

File: src/test/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineNodeUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    List<FlowNodeWrapper> stepNodes = builder.getStageSteps(stageId);
+    FlowNodeWrapper errorStep = stepNodes.get(0);
+    assertThat(PipelineNodeUtil.getExceptionText(errorStep.getNode()), is("This is an error"));
+  }
+
+  @Issue("GH#213")
+  @Test
+  public void pipelineCallsUndefinedVar() throws Exception {
+    // It's a bit dirty, but do this in one to avoid reloading and rerunning the job (as it takes a
+    // long time)
```
Comment (timja) [inline src/test/java/io/jenkins/plugins/pipelinegraphview/utils/PipelineNodeUtilTest.java:52]: is this comment needed? unsure how it relates to the test it's just running a build once and testing one thing?
```suggestion
```

File: src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java [hunk 3]
```diff
@@ -55,35 +55,44 @@ public String getIconClassName() {
...
-        logger.debug("Steps: '" + stepsJson + "'.");
-      }
-      return HttpResponses.okJSON(JSONObject.fromObject(stepsJson));
+      return HttpResponses.okJSON(getSteps(nodeId));
     } else {
       return HttpResponses.errorJSON("Error getting console text");
     }
   }
 
+  // Private method for testing.
```
Comment (timja) [inline src/main/java/io/jenkins/plugins/pipelinegraphview/consoleview/PipelineConsoleViewAction.java:64]: the method is protected? am I missing something I don't see it getting called in a test?

**Comment by timja** [comment]
> It would be nice to find a way to make the tests quicker (e.g. not needing a Jenkins job to run). Does anyone know if there's a way to serialise a run?

Could probably zip up a job run but would make maintaining it a pain I think, the jobs themselves should be quite quick to run.

Maybe post on the developers mailing list?

**Comment by timja** [comment]
Pushed some tidyup and tested it locally

**Comment by timja** [comment]
> It would be nice to find a way to make the tests quicker (e.g. not needing a Jenkins job to run). Does anyone know if there's a way to serialise a run?
> 
> I feel there is a better way to get steps in the tests (shouldn't rely on PipelineStepVisitor as if it breaks a lot of other tests break). I tried with `getNodesByDisplayName` - which uses the upstream `FlowGraphTable`, but couldn't find the correct way to get a step (works fine for stages).

Full build including tests takes 01:44 min for me locally which isn't bad (but I'm sure we can improve it 😄 )

**Comment by timja** [comment]
> It would be nice to find a way to make the tests quicker

Although it takes over 3x slower on windows 😢 https://github.com/jenkins-infra/helpdesk/issues/3117

**Comment by timbrown5** [comment]
Thanks for the help. I wouldn't have have been able to look at this for a few days 👍

**Pull request merged**